### PR TITLE
bpf: Introduce per-cluster conntrack maps

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
This is a part of the Cluster Mesh with overlapping PodCIDR support. When the remote cluster has an overlapping PodCIDR, we must identify the remote endpoint using IP + ClusterID. CT is not an exception, so it's ideal for extending the CT map's key with a new ClusterID field. However, the problem is we cannot extend the CT map without breaking users' connections now. Thus, we created a new map-in-map containing the CT maps for each cluster.

This PR only contains basic map definitions and tests. No one uses it. We also don't have GC for per-cluster maps because GC depends on the NAT map. Once I implement per-cluster NAT (it has the same problem), I'll publish another PR for GC.

```release-note
bpf: Introduce per-cluster conntrack maps
```

### CODE REVIEW

File: pkg/maps/ctmap/per_cluster_ctmap.go [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+// remote clusters. It is implemented as an array-of-maps which
+// its index is a ClusterID.
+//
+// Why can't we use global CT maps? That's because we currently
+// don't have a good way of extending CT map's key without breaking
+// user's connection. Thus, instead of extending existing CT map
+// key with ClusterID, we chose to create CT map per-cluster. When
+// we have a good way of extending global CT maps in the future, we
+// should retire this entire file.
+type PerClusterCtMap struct {
```
Comment (ti-mo) [inline pkg/maps/ctmap/per_cluster_ctmap.go]: `PerClusterCTMap`? https://go.dev/talks/2014

File: pkg/maps/ctmap/per_cluster_ctmap.go [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+// and privileged tests or dummyPerClusterCtMaps for unprivileged
+// testing.
+var PerClusterCtMaps PerClusterCtMapsIface
+
+const (
+	perClusterCtMapMaxEntries = cmtypes.ClusterIDMax + 1
+)
+
+// An interface to interact with all per-cluster CT maps
+type PerClusterCtMapsIface interface {
```
Comment (ti-mo) [inline pkg/maps/ctmap/per_cluster_ctmap.go]: Following the Go convention: `PerClusterCTMapper`

File: pkg/maps/ctmap/per_cluster_ctmap_privileged_test.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (ti-mo) [inline pkg/maps/ctmap/per_cluster_ctmap_privileged_test.go]: Feel free to rename this to `per_cluster_ctmap_test.go`. With the removal of the `privileged_tests` build tag, privileged tests no longer need to be split into separate files, and the filename is getting rather long.

File: pkg/maps/ctmap/per_cluster_ctmap_privileged_test.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+func (k *PerClusterCtMapPrivilegedTestSuite) SetUpSuite(c *C) {
+	testutils.PrivilegedCheck(c)
+
+	bpf.CheckOrMountFS("")
+	err := rlimit.RemoveMemlock()
+	c.Assert(err, IsNil)
+
+	// Per-cluster CT map depends on the map information
+	InitMapInfo(option.CTMapEntriesGlobalTCPDefault, option.CTMapEntriesGlobalAnyDefault, true, true, true)
```
Comment (ti-mo) [inline pkg/maps/ctmap/per_cluster_ctmap_privileged_test.go]: This is already called in `ctmap_privileged_test.go:init()`.

**Review by ti-mo — CHANGES_REQUESTED** [review]

File: pkg/clustermesh/types/types.go [hunk 5]
```diff
@@ -15,6 +15,18 @@ const (
 	ClusterIDMax = 255
 )
 
+func ValidateClusterID(clusterID uint32) error {
```
Comment (sayboras) [inline pkg/clustermesh/types/types.go:18]: is this change already part of https://github.com/cilium/cilium/pull/22875 ?

Comment (YutaroHayakawa) [reply pkg/clustermesh/types/types.go:18]: Yes, but both PRs depend on this change, so let me rebase the branch after the one is merged into the master.

**Comment by YutaroHayakawa** [comment]
Applied the fix suggested by Timo.

**Review by ti-mo — APPROVED** [review]

**Review by NikAleksandrov — APPROVED** [review]

**Comment by YutaroHayakawa** [comment]
- k8s-1.16-kernel-4.9 is failing because map-in-map is from 4.12

**Comment by YutaroHayakawa** [comment]
I surrounded the per-cluster CT maps with guard macro `ENABLE_CLUSTER_AWARE_ADDRESSING`.

https://github.com/cilium/cilium/pull/22857/files#diff-5c024a095befbdabbdd926da1b0c8937d6dec93f4b7f7f09cfcf431b73bed814R40

And added new maps to ignore list.

https://github.com/cilium/cilium/pull/22857/files#diff-24b6dbe6806c91a53adf84e5aa545c62de4474c19e276e1ff2bc6e8207273143R65

**Pull request merged**